ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -192,17 +269,137 @@ static INT_PTR CALLBACK ShowMessageExtProc(HWND hwndDlg, UINT uMsg, WPARAM wPara
   return 0;
 }
 
+/* < Used by GetDirectoryProc > */
+WCHAR *LowerCaseToActualPathName(WCHAR *wstr_dname)
+{
+  LPITEMIDLIST pstr_dname;
+  SHParseDisplayName(wstr_dname, 0, &pstr_dname, 0, 0);
+
+  static WCHAR wstr_result[MAX_PATH + 1];
+  SHGetPathFromIDListW(pstr_dname, wstr_result);
+  return wstr_result;
+}
+
+WCHAR wstr_dname[MAX_PATH + 1];
+WCHAR wstr_stc1[MAX_PATH + 1];
+static string DlgItemText;
+tstring ActualPath;
+/* < / Used by GetDirectoryProc > */
+
+UINT APIENTRY GetDirectoryProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+  switch (uMsg)
+  {
+  case WM_INITDIALOG:
+    {
+      ClientResize(hWnd, 424, 255);
+      RECT rect;
+      GetWindowRect(hWnd, &rect);
+      MoveWindow(hWnd,
+        (GetSystemMetrics(SM_CXSCREEN) / 2) - ((rect.right - rect.left) / 2),
+        (GetSystemMetrics(SM_CYSCREEN) / 2) - ((rect.bottom - rect.top) / 2),
+        rect.right - rect.left, rect.bottom - rect.top, TRUE);
+      MoveWindow(GetDlgItem(hWnd, IDOK), 256, 224, 77, 27, TRUE);
+      MoveWindow(GetDlgItem(hWnd, IDCANCEL), 340, 224, 77, 27, TRUE);
+      MoveWindow(GetDlgItem(hWnd, stc3), 232, 56, 72, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, IDC_LABEL), 8, 56, 72, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, stc2), 8, 8, 93, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, stc4), 232, 176, 50, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, stc1), 8, 24, 409 * 100, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, IDC_DNAME), 8, 24, 409, 16, TRUE);
+      MoveWindow(GetDlgItem(hWnd, lst1), 232, 72, 185, 93, TRUE);
+      MoveWindow(GetDlgItem(hWnd, lst2), 8, 72, 213, 123, TRUE);
+      MoveWindow(GetDlgItem(hWnd, cmb2), 232, 192, 185, 19, TRUE);
```

Comment:
Why are we doing this here instead of the rc file? This is the entire point of the RC file. Doing it this way adds unnecessary time to construct and initialize the dialog whereas the rc file compiles the coordinates and will immediately construct the controls in the right location. This also seems to be inconsistent with how we've placed controls for every other dialog in these sources. If possible, please convert this and follow the other conventions already established in here.

Reply:
because RC files are not based on pixels - they are based on dialog coordinates. The dialogs I am imitating are pixel-exact. I do not know how to convert pixels to dialog coordinates.

Reply:
If i knew how to convert the measurements I would prefer to do it that way just like you do

Reply:
In addition, I think it is stupid personally to follow the way you guys originally had the buttons place because that is a non-standard that no one in the Windows software industry goes by. The get_string, get_password, get_integer, and get_password are designed to look EXACTLY like the standard VBScript Input box. We had a whole discussion about this on discord and you are acting like you don't remember any of it. There are actually a crap ton of things here that we've already been over about a month ago and here I am repeating myself about stuff I thought we already had established. Also get_directory looks exactly like GM's get_directory. It was moddled pixel-exact after the standard Delphi-based select directory dialog.

Reply:
I was planning to redesign the last two remaining functions, (show_message_ext and get_login), to follow some kind of standard, since those are the only ones I haven't already done that with.

Reply:
Let me break down your comments here and try to reply.

1. I see what you are saying about the dialog coordinates, I'll try to find some info to help.
2. I do remember having the conversation about imitating the VB input dialog but you're conflating issues. There's a huge issue you are not considering here, that original dialog in VB lacked (and likely still lacks) DPI scaling. I mean you should be trying to imitate only the good parts of it (not the bad parts like caring if it's "pixel" perfect instead of properly scaled to the user's monitor). The point here is that you can take what VB already had great, and make it even better. Don't take what it had and make it worse.
3. I don't understand the context of this comment, non-standard... by what standard? Are you saying our `show_message_ext` or `get_login` is incompatible in some way with GM? What were you planning to change?
4. Why do you always have to go to the extreme? 😋 You've done some indisputably GOOD things here, like UTF-8 support and have cleaned up some code in positive ways. This just happens to be a lot to review. Maybe I should mention too that this is usually why I try to focus my own changesets around the concept of "indisputably" better. To make my life easier and myself less tried I always try to do changes in a way that nobody can say I didn't something bad, you should maybe try that yourself as well. It requires a bit of clairvoyance, thinking ahead like "ok, before I send this code, what is something I can see Josh or Robert picking about?" I understand that question may be hard for you to answer, because you still haven't interacted with us much on this tracker yet. But regardless, at the end of the day, you know you tried and at the very least I can say thanks Sam for at least trying to get these widgets cleaned up.

I'll get you some info and you can respond to these comments, we'll go from there.

Reply:
2. That's good then, it wasn't clear to me. In a way I was using that as an example to explain why you don't have to be 100% like VB6 in every way. I hope you see why.
3. I have to just slightly dispute one thing here, because in GM8.1 and lower it is sort of a custom dialog because you can use the `message_box_*` family of functions to change the texture. Even `show_message` in older GM is much more custom than native. That's ok though, we never really intended to support those `message_box_*`  functions anyway. That's a good idea for `get_login`, but save it for another time/pr just to make this easier for us all.
4. You're welcome 👍 

Finally, I want to restate two things.
1) I would still favor having the RC file so that the dialog is compiled and less likely to flicker upon construction/load faster. (Just because you don't see flickering doesn't mean it can't manifest later in some other way!) I will actually look for some info to help with this now, I had forgotten to come back to this comment earlier.
2) Despite everything I've said, when it comes to making an actual game, I actually side with YoYoGame's when the manual states the following:

> NOTE: THIS FUNCTION IS FOR DEBUG USE ONLY. Should you require this functionality in your final game, please use get_integer_async

I know they aren't saying to use a custom UI, but I think they are trying to hint at you to do that here. The manual also states that just `show_message` is for debug use only.

I really think a game looks best when it uses its own custom dialogs themed to match the game. Using native dialogs in a game can detract from the entertainment experience depending on the game sometimes in my opinion. Certain types of games, like an RTS, I would personally be more lenient if they showed me a native dialog. Now if it comes to a file chooser, I'm almost always going to favor the native one, even if it's in a game.